### ISSUE TITLE
Bug cause Transforms/test-legalizer.mlir failed, it isn't caught by CI because of it's a assert. 

### ISSUE THREAD

**Issue by Johnson9009** [tensorflow/mlir#32]
The latest commits will cause test `Transforms/test-legalizer.mlir` failed when build project as debug  mode.

The failure is caught by an assert of llvm `assert(!NodePtr->isKnownSentinel());`, I guess the CI are all release build, so this failure is not caught.

The root cause of this issue is below new added code of file `test/TestDialect/TestPatterns.cpp` added by commit: 1fc814032a35bd19e5aedc68ab7d480450d4b386
```cpp
    // Inline this region into the parent region.
    auto &parentRegion = *op->getContainingRegion();
    rewriter.inlineRegionBefore(op->getRegion(0), parentRegion.end());
```
`parentRegion.end()` give the end iterator, from below code we can found it is passed to `before`, but `before` need to be an iterator which point to available item, because it need to call the function `getParent()`.

```cpp
void PatternRewriter::inlineRegionBefore(Region &region,
                                         Region::iterator before) {
  before->getParent()->getBlocks().splice(before, region.getBlocks());
}
```
https://github.com/tensorflow/mlir/blob/960b39c8773d4fdf9d8d2291680922293617be77/test/TestDialect/TestPatterns.cpp#L63
https://github.com/tensorflow/mlir/blob/960b39c8773d4fdf9d8d2291680922293617be77/lib/IR/PatternMatch.cpp#L124

The Regression Test Failure Report:
```
[0/1] Running the MLIR regression tests
-- Testing: 208 tests, 8 threads --
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 
FAIL: MLIR :: Transforms/test-legalizer.mlir (190 of 208)
******************** TEST 'MLIR :: Transforms/test-legalizer.mlir' FAILED ********************
Script:
--
: 'RUN: at line 1';   /u/qiangz/workspaces/llvm-project/build/bin/mlir-test-opt -test-legalize-patterns -split-input-file -verify-diagnostics /remote/ailab1/qiangz/workspaces/llvm-project/llvm/projects/mlir/test/Transforms/test-legalizer.mlir | /u/qiangz/workspaces/llvm-project/build/bin/FileCheck /remote/ailab1/qiangz/workspaces/llvm-project/llvm/projects/mlir/test/Transforms/test-legalizer.mlir --dump-input=fail
--
Exit Code: 2

Command Output (stderr):
--
mlir-test-opt: /u/qiangz/workspaces/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:139: llvm::ilist_iterator::reference llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void>, false, false>::operator*() const [OptionsT = llvm::ilist_detail::node_options<mlir::Block, true, false, void>, IsReverse = false, IsConst = false]: Assertion `!NodePtr->isKnownSentinel()' failed.
Stack dump:
0.	Program arguments: /u/qiangz/workspaces/llvm-project/build/bin/mlir-test-opt -test-legalize-patterns -split-input-file -verify-diagnostics /remote/ailab1/qiangz/workspaces/llvm-project/llvm/projects/mlir/test/Transforms/test-legalizer.mlir 
1.	Program arguments: /u/qiangz/workspaces/llvm-project/build/bin/mlir-test-opt -test-legalize-patterns -split-input-file -verify-diagnostics /remote/ailab1/qiangz/workspaces/llvm-project/llvm/projects/mlir/test/Transforms/test-legalizer.mlir 
 #0 0x0000000000544e29 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /u/qiangz/workspaces/llvm-project/llvm/lib/Support/Unix/Signals.inc:494:11
 #1 0x0000000000544fd9 PrintStackTraceSignalHandler(void*) /u/qiangz/workspaces/llvm-project/llvm/lib/Support/Unix/Signals.inc:558:1
 #2 0x00000000005438e6 llvm::sys::RunSignalHandlers() /u/qiangz/workspaces/llvm-project/llvm/lib/Support/Signals.cpp:67:5
 #3 0x000000000054565b SignalHandler(int) /u/qiangz/workspaces/llvm-project/llvm/lib/Support/Unix/Signals.inc:357:1
 #4 0x00007f30414795e0 __restore_rt (/lib64/libpthread.so.0+0xf5e0)
 #5 0x00007f303ffeb1f7 __GI_raise (/lib64/libc.so.6+0x351f7)
 #6 0x00007f303ffec8e8 __GI_abort (/lib64/libc.so.6+0x368e8)
 #7 0x00007f303ffe4266 __assert_fail_base (/lib64/libc.so.6+0x2e266)
 #8 0x00007f303ffe4312 (/lib64/libc.so.6+0x2e312)
 #9 0x000000000057acbb llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void>, false, false>::operator*() const /u/qiangz/workspaces/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:140:33
#10 0x00000000006ac435 llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void>, false, false>::operator->() const /u/qiangz/workspaces/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:142:32
#11 0x00000000006aa3fd mlir::PatternRewriter::inlineRegionBefore(mlir::Region&, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void>, false, false>) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/IR/PatternMatch.cpp:124:11
#12 0x0000000000590a0d (anonymous namespace)::DialectConversionRewriter::inlineRegionBefore(mlir::Region&, llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Block, true, false, void>, false, false>) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:432:3
#13 0x00000000004a24d1 (anonymous namespace)::TestRegionRewriteBlockMovement::matchAndRewrite(mlir::Operation*, llvm::ArrayRef<mlir::Value*>, mlir::PatternRewriter&) const /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/test/TestDialect/TestPatterns.cpp:66:5
#14 0x000000000058d5ab mlir::ConversionPattern::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:489:12
#15 0x00000000005924b5 (anonymous namespace)::OperationLegalizer::legalizePattern(mlir::Operation*, mlir::RewritePattern*, (anonymous namespace)::DialectConversionRewriter&) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:636:8
#16 0x00000000005921f5 (anonymous namespace)::OperationLegalizer::legalize(mlir::Operation*, (anonymous namespace)::DialectConversionRewriter&) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:601:19
#17 0x0000000000591dbd (anonymous namespace)::FunctionConverter::convertBlock((anonymous namespace)::DialectConversionRewriter&, mlir::Block*, llvm::DenseSet<mlir::Block*, llvm::DenseMapInfo<mlir::Block*> >&) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:853:23
#18 0x00000000005903b9 (anonymous namespace)::FunctionConverter::convertRegion((anonymous namespace)::DialectConversionRewriter&, mlir::Region&, bool) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:890:14
#19 0x000000000058e9e3 (anonymous namespace)::FunctionConverter::convertFunction(mlir::Function*, mlir::TypeConverter::SignatureConversion*) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:917:11
#20 0x000000000058e7f3 mlir::applyConversionPatterns(llvm::ArrayRef<mlir::Function*>, mlir::ConversionTarget&, mlir::TypeConverter&, std::vector<std::unique_ptr<mlir::RewritePattern, std::default_delete<mlir::RewritePattern> >, std::allocator<std::unique_ptr<mlir::RewritePattern, std::default_delete<mlir::RewritePattern> > > >&&) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:1126:30
#21 0x000000000058e57b mlir::applyConversionPatterns(mlir::Module&, mlir::ConversionTarget&, mlir::TypeConverter&, std::vector<std::unique_ptr<mlir::RewritePattern, std::default_delete<mlir::RewritePattern> >, std::allocator<std::unique_ptr<mlir::RewritePattern, std::default_delete<mlir::RewritePattern> > > >&&) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Transforms/DialectConversion.cpp:1097:10
#22 0x00000000004a0cdb (anonymous namespace)::TestLegalizePatternDriver::runOnModule() /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/test/TestDialect/TestPatterns.cpp:113:16
#23 0x00000000005b8007 mlir::ModulePassBase::run(mlir::Module*, mlir::ModuleAnalysisManager&) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Pass/Pass.cpp:92:3
#24 0x00000000005b83ea mlir::detail::ModulePassExecutor::run(mlir::Module*, mlir::ModuleAnalysisManager&) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Pass/Pass.cpp:132:22
#25 0x00000000005b8c62 mlir::PassManager::run(mlir::Module*) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/lib/Pass/Pass.cpp:267:15
#26 0x0000000000467235 performActions(llvm::SourceMgr&, mlir::MLIRContext*) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/tools/mlir-opt/mlir-opt.cpp:96:17
#27 0x00000000004670c5 processFile(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/tools/mlir-opt/mlir-opt.cpp:138:34
#28 0x0000000000466f2f splitAndProcessFile(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >) /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/tools/mlir-opt/mlir-opt.cpp:166:9
#29 0x0000000000466bcd main /u/qiangz/workspaces/llvm-project/llvm/projects/mlir/tools/mlir-opt/mlir-opt.cpp:197:5
#30 0x00007f303ffd7c05 __libc_start_main (/lib64/libc.so.6+0x21c05)
#31 0x0000000000407a69 _start (/u/qiangz/workspaces/llvm-project/build/bin/mlir-test-opt+0x407a69)
FileCheck error: '-' is empty.
FileCheck command line:  /u/qiangz/workspaces/llvm-project/build/bin/FileCheck /remote/ailab1/qiangz/workspaces/llvm-project/llvm/projects/mlir/test/Transforms/test-legalizer.mlir --dump-input=fail

--

********************
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 
Testing Time: 3.02s
********************
Failing Tests (1):
    MLIR :: Transforms/test-legalizer.mlir

  Expected Passes    : 207
  Unexpected Failures: 1
FAILED: projects/mlir/test/CMakeFiles/check-mlir 
cd /u/qiangz/workspaces/llvm-project/build/projects/mlir/test && /usr/bin/python2.7 /u/qiangz/workspaces/llvm-project/build/./bin/llvm-lit -sv /u/qiangz/workspaces/llvm-project/build/projects/mlir/test
ninja: build stopped: subcommand failed.
```

**Comment by River707** [comment]
Thanks, should be fixed in the next sync.

**Comment by Johnson9009** [comment]
This issue is fixed by commit: 262adda9ed4a969017c63a6aec4e7372927b8714. @River707 Thanks.
